src/main/scala/org/bitcoins/core/script/crypto/HashType.scala

```diff
@@ -1,47 +1,145 @@
+  }
+
+  def fromNumber(num : Int32) : HashType = {
+    if (isSIGHASH_NONE(num)) {
+      if (isSIGHASH_NONE_ANYONECANPAY(num)) SIGHASH_NONE_ANYONECANPAY(num) else SIGHASH_NONE(num)
+    }
+    else if (isSIGHASH_SINGLE(num)) {
+      if (isSIGHASH_SINGLE_ANYONECANPAY(num)) SIGHASH_SINGLE_ANYONECANPAY(num) else SIGHASH_SINGLE(num)
+    }
+    else if (isSIGHASH_ANYONECANPAY(num)) {
+      if (isSIGHASH_ALL_ANYONECANPAY(num)) SIGHASH_ALL_ANYONECANPAY(num) else SIGHASH_ANYONECANPAY(num)
+    }
+    else {
+      SIGHASH_ALL(num)
+    }
+  }
+
+  def byte (hashType : HashType) : Byte = hashType match {
+    case zero if zero.num == Int32.zero => 0x00.toByte
+    case _ : SIGHASH_ALL => 0x01.toByte
+    case _ : SIGHASH_NONE => 0x02.toByte
+    case _ : SIGHASH_SINGLE => 0x03.toByte
+    case _ : SIGHASH_ANYONECANPAY => 0x80.toByte
+    case _ : SIGHASH_ALL_ANYONECANPAY => 0x81.toByte
+    case _ : SIGHASH_NONE_ANYONECANPAY => 0x82.toByte
+    case _ : SIGHASH_SINGLE_ANYONECANPAY => 0x83.toByte
+  }
+
+  def isSIGHASH_ALL_ONE(num : Int32) : Boolean = (num & Int32(0x1f)) == Int32(1)
+  def isSIGHASH_NONE(num : Int32) : Boolean = (num & Int32(0x1f)) == Int32(2)
+  def isSIGHASH_SINGLE(num : Int32) : Boolean = (num & Int32(0x1f)) == Int32(3)
+  def isSIGHASH_ANYONECANPAY(num : Int32) : Boolean = (num & Int32(0x80)) == Int32(0x80)
+  def isSIGHASH_ALL_ANYONECANPAY(num : Int32) : Boolean = isSIGHASH_ALL_ONE(num) && isSIGHASH_ANYONECANPAY(num)
+  def isSIGHASH_NONE_ANYONECANPAY(num : Int32) : Boolean = isSIGHASH_NONE(num) && isSIGHASH_ANYONECANPAY(num)
+  def isSIGHASH_SINGLE_ANYONECANPAY(num: Int32) : Boolean = isSIGHASH_SINGLE(num) && isSIGHASH_ANYONECANPAY(num)
+  def isSIGHASH_ALL(num : Int32) : Boolean = {
+    if (!(isSIGHASH_NONE(num) || isSIGHASH_SINGLE(num) || isSIGHASH_ANYONECANPAY(num) || isSIGHASH_ALL_ANYONECANPAY(num) ||
+      isSIGHASH_SINGLE_ANYONECANPAY(num) || isSIGHASH_NONE_ANYONECANPAY(num))) true
+    else false
+  }
+
+  val hashTypes = Seq(SIGHASH_ALL.value, SIGHASH_NONE, SIGHASH_SINGLE, SIGHASH_ANYONECANPAY,
+    SIGHASH_NONE_ANYONECANPAY, SIGHASH_ALL_ANYONECANPAY, SIGHASH_SINGLE_ANYONECANPAY)
+
+}
+
+object HashTypeOperations extends HashTypeOperations
+
 /**
  * This seems strange, but it needs to take a parameter. This is because
  * SIGHASH_ALL is essentially a catch all if the none of the other hash types are matched.
  * Therefore SIGHASH_ALL could be represented by the byte 0x05 since 0x05 does not match
  * any of the other hash types. The default byte for SIGHASH_ALL is 0x01
  *
- * @param byte
  */
-case class SIGHASH_ALL(byte : Byte = 0x01.toByte) extends HashType
 
-case object SIGHASH_NONE extends HashType {
-  override def byte = 0x02.toByte
+ object SIGHASH_ALL extends Factory[HashType] with HashTypeOperations {
+  private case class SIGHASH_ALLImpl(num: Int32) extends SIGHASH_ALL {
+    require(isSIGHASH_ALL(num), "SIGHASH_ALL acts as a 'catch-all' for undefined hashtypes, as well as the values of zero and one.")
```

Comment:
This isn't really a useful error message, if I get an error I want to know what the original input was, and why it failed. I can't even see what `num` was in the error message to cause it to fail